internal/oci/runtime_oci.go

```diff
@@ -906,7 +908,7 @@ func (r *runtimeOCI) StopLoopForContainer(c *Container, bm kwait.BackoffManager)
 			select {
 			case <-ctx.Done():
 				return
-			case <-time.After(stopProcessWatchSleep):
```

Comment:
Why? The end result is exactly the same. Have you seen some issues?

Reply:
Also, I am glad you are looking in this direction. I wanted to ask about this function. We have it and the other, inner-loop, invocation of `Living()`, and it seems it's either a race or these aren't stopping the loop to stop the spurious kills. Hmm.

Reply:
I think the new logic eliminates the need for error string matching by centralizing the c.Living() check within a ticker-based goroutine, which reduces the likelihood of race conditions.

Reply:
> Why? The end result is exactly the same. Have you seen some issues?

I didn't find any issues. I just thought it was more idiomatic.

Reply:
> I think the new logic eliminates the need for error string matching by centralizing the c.Living() check within a ticker-based goroutine, which reduces the likelihood of race conditions.

@sohankunkerkar, that is much better, indeed!

Where we would still have issues would be the kill loop if the timeout is set to 0. However, this can only really be set when talking to the CRI directly, so with crictl, for example.

Curious that I couldn't set a timeout for termination grace period lower than 2 seconds. Haven't looked into why.

Reply:
>Where we would still have issues would be the kill loop if the timeout is set to 0. However, this can only really be set when talking to the CRI directly, so with crictl, for example.

I tried running the following example locally (a modified version of `ctr multiple stop calls`), and it worked without any problem.
```sh
@test "ctr stop with zero timeout" {
    start_crio

    # Create a container with a long-running command to simulate a scenario where
    # a container takes a while to stop gracefully.
    jq '.command = ["/bin/sh", "-c", "sleep 600"]' \
        "$TESTDATA"/container_config.json > "$newconfig"
    ctr_id=$(crictl run "$newconfig" "$TESTDATA"/sandbox_config.json)

    # Issue the crictl stop command with a timeout of 0.
    crictl stop --timeout 0 "$ctr_id" &> /dev/null

    # Verify that the container has either stopped or exited.
    final_state=$(crictl inspect "$ctr_id" | grep -Po '(?<="state": ")[^"]*')
    if [ "$final_state" != "CONTAINER_STOPPED" ] && [ "$final_state" != "CONTAINER_EXITED" ]; then
        echo "Test failed: Container did not stop or exit as expected."
        exit 1
    fi
}
```

Reply:
>Curious that I couldn't set a timeout for termination grace period lower than 2 seconds. Haven't looked into why.

I used this same example to set the `timeout` value as `1` and it worked.

Reply:
> > Curious that I couldn't set a timeout for termination grace period lower than 2 seconds. Haven't looked into why.
> 
> I used this same example to set the `timeout` value as `1` and it worked.

Sorry, I should have mentioned `terminationGracePeriodSeconds` from the Pod object specification rather than just the "grace period".

Reply:
> > Where we would still have issues would be the kill loop if the timeout is set to 0. However, this can only really be set when talking to the CRI directly, so with crictl, for example.
> 
> I tried running the following example locally (a modified version of `ctr multiple stop calls`), and it worked without any problem.
[...]

I am glad it works! However, we were trying to suppress spurious logs here in the Pull Request initially. When you set timeout to 0, that does not happen, per:

```console
root@worker-node01:~# crictl run container.json sandbox.json 
810eb62698d7bab94f923b144fcd915551ad2de30e60248cefb76e6c9c66506b
```

```console
root@worker-node01:~# crictl ps
CONTAINER           IMAGE                                CREATED             STATE               NAME                ATTEMPT             POD ID              POD
810eb62698d7b       quay.io/crio/fedora-crio-ci:latest   10 seconds ago      Running             busybox             0                   554ce2b4a1fd1       unknown
root@worker-node01:~# 
```

```console
root@worker-node01:~# crictl stop --timeout=0 810eb62698d7b
810eb62698d7b
root@worker-node01:~# echo $?
0
```

``console
root@worker-node01:~# crictl inspect 810eb62698d7b | grep -Po '(?<="state": ")[^"]*'
CONTAINER_EXITED
```

```console
INFO[2024-08-05 15:13:36.229891070Z] Stopping container: 810eb62698d7b (timeout: 0s)  id=4a7b454a-739c-4881-bdce-f4fb9c8619e4 name=/runtime.v1.RuntimeService/StopContainer
WARN[2024-08-05 15:13:36.235362788Z] Stopping container 810eb62698d7bab94f923b144fcd915551ad2de30e60248cefb76e6c9c66506b with stop signal timed out. Killing... 
WARN[2024-08-05 15:13:36.235403944Z] XXX 1 810eb62698d7bab94f923b144fcd915551ad2de30e60248cefb76e6c9c66506b 
ERRO[2024-08-05 15:13:36.239394660Z] Killing container 810eb62698d7bab94f923b144fcd915551ad2de30e60248cefb76e6c9c66506b failed: `/usr/bin/crio-runc --root /run/runc --systemd-cgroup kill 810eb62698d7bab94f923b144fcd915551ad2de30e60248cefb76e6c9c66506b KILL` failed: time="2024-08-05T15:13:36Z" level=error msg="container not running"
 : exit status 1 
INFO[2024-08-05 15:13:36.249060431Z] Stopped container 810eb62698d7bab94f923b144fcd915551ad2de30e60248cefb76e6c9c66506b: //  id=4a7b454a-739c-4881-bdce-f4fb9c8619e4 name=/runtime.v1.RuntimeService/StopContainer
```

This is what I meant. The `CONTAINER_EXITED` status is most likely set for a container that terminated.

Reply:
> > > Curious that I couldn't set a timeout for termination grace period lower than 2 seconds. Haven't looked into why.
> > 
> > 
> > I used this same example to set the `timeout` value as `1` and it worked.
> 
> Sorry, I should have mentioned `terminationGracePeriodSeconds` from the Pod object specification rather than just the "grace period".

An example:

```console
$ k get pod test-3 -o yaml | grep terminationGracePeriodSeconds:
  terminationGracePeriodSeconds: 1
```

```console
INFO[2024-08-05 15:37:15.109684729Z] Stopping container: f9ef70b73d1e7f0085ca69497ce290b7306a7b7061b9173335ec91d9a9566758 (timeout: 2s)  id=b90d2f54-8419-4241-ac22-e22fc760bc64 name=/runtime.v1.RuntimeService/StopContainer
INFO[2024-08-05 15:37:15.223443674Z] Stopped container f9ef70b73d1e7f0085ca69497ce290b7306a7b7061b9173335ec91d9a9566758: default/test-3/test-3  id=b90d2f54-8419-4241-ac22-e22fc760bc64 name=/runtime.v1.RuntimeService/StopContainer
INFO[2024-08-05 15:37:15.223824002Z] Stopping pod sandbox: a4c0fa4d3043a2e7f94359cfc5676658181226328f4627b42d845c1f24c26426  id=6b14eb06-2442-4f0a-b1b4-f340fe1437a6 name=/runtime.v1.RuntimeService/StopPodSandbox
INFO[2024-08-05 15:37:15.223977168Z] Got pod network &{Name:test-3 Namespace:default ID:a4c0fa4d3043a2e7f94359cfc5676658181226328f4627b42d845c1f24c26426 UID:550731f3-136d-412b-b684-a1ee7b85c0c4 NetNS:/var/run/netns/034fa416-0c1f-442c-81d5-f2237ac83c14 Networks:[{Name:crio Ifname:eth0}] RuntimeConfig:map[crio:{IP: MAC: PortMappings:[] Bandwidth:<nil> IpRanges:[] CgroupPath:kubepods-besteffort-pod550731f3_136d_412b_b684_a1ee7b85c0c4.slice PodAnnotations:0xc00014a560}] Aliases:map[]} 
INFO[2024-08-05 15:37:15.224052221Z] Deleting pod default_test-3 from CNI network "crio" (type=bridge) 
INFO[2024-08-05 15:37:15.311514819Z] Stopped pod sandbox: a4c0fa4d3043a2e7f94359cfc5676658181226328f4627b42d845c1f24c26426  id=6b14eb06-2442-4f0a-b1b4-f340fe1437a6 name=/runtime.v1.RuntimeService/StopPodSandbox
INFO[2024-08-05 15:37:16.207227134Z] Removing container: f9ef70b73d1e7f0085ca69497ce290b7306a7b7061b9173335ec91d9a9566758  id=3b6147f7-dc59-48ac-9cde-289f3da8d9f9 name=/runtime.v1.RuntimeService/RemoveContainer
INFO[2024-08-05 15:37:16.214599391Z] Removed container f9ef70b73d1e7f0085ca69497ce290b7306a7b7061b9173335ec91d9a9566758: default/test-3/test-3  id=3b6147f7-dc59-48ac-9cde-289f3da8d9f9 name=/runtime.v1.RuntimeService/RemoveContainer
```

I haven't looked why that is, to be completely honest. Albeit, very curious, indeed.

Reply:
@kwilczynski Thanks for the thorough examples. Let me re-visit this logic.

Reply:
I think for `terminationGracePeriodSeconds`, some of the pieces are covered in https://github.com/kubernetes/kubernetes/issues/83916

Reply:
>I am glad it works! However, we were trying to suppress spurious logs here in the Pull Request initially. When you set timeout to 0, that does not happen, per:

I'm leaning towards the combination of the existing logic + the error matching that we discussed earlier, thoughts?

Reply:
@sohankunkerkar, that would be fine, indeed.

Reply:
> I think for `terminationGracePeriodSeconds`, some of the pieces are covered in [kubernetes/kubernetes#83916](https://github.com/kubernetes/kubernetes/issues/83916)

@sohankunkerkar, thank you for this!

That looks interesting. Not quite the same thing, since for the simple example, I wasn't updating the value. Also, this is something that we should support in CRI-O correctly, too. Hopefully we are. :smile: